src/main/java/com/shapesecurity/shift/es2016/semantics/visitor/ConstantFolder.java

```diff
@@ -0,0 +1,141 @@
+import com.shapesecurity.shift.es2016.semantics.asg.Node;
+import com.shapesecurity.shift.es2016.semantics.asg.NodeWithValue;
+import com.shapesecurity.shift.es2016.semantics.asg.UnaryOperation.Not;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+
+import static com.shapesecurity.shift.es2016.semantics.visitor.ECMA262Operations.ToInt32;
+import static com.shapesecurity.shift.es2016.semantics.visitor.ECMA262Operations.ToNumber;
+import static com.shapesecurity.shift.es2016.semantics.visitor.ECMA262Operations.ToString;
+import static com.shapesecurity.shift.es2016.semantics.visitor.ECMA262Operations.ToUint32;
+import static com.shapesecurity.shift.es2016.semantics.visitor.ECMA262Operations.Type;
+import static com.shapesecurity.shift.es2016.semantics.visitor.Truthiness.truthiness;
+
+public class ConstantFolder extends ReconstructingReducer {
+    private final Semantics semantics;
+
+    public ConstantFolder(Semantics semantics) {
+        this.semantics = semantics;
+    }
+
+    @Nonnull
+    public static Semantics reduce(Semantics input) {
+        return new Semantics(new ConstantFolder(input).go(), input.locals, input.scriptVarDecls);
+    }
+
+    @Nonnull
+    private Node go() {
+        return this.visitNode(this.semantics.node);
+    }
+
+    @Nonnull
+    @Override
+    protected NodeWithValue visitNot(@Nonnull Not not) {
+        not = (Not) super.visitNot(not);
+        if (not.expression instanceof LiteralNumber) {
+            LiteralNumber n = (LiteralNumber) not.expression;
+            return new LiteralBoolean(n.value == 0);
+        } else if (not.expression instanceof LiteralBoolean) {
+            return new LiteralBoolean(!((LiteralBoolean) not.expression).value);
+        } else if (not.expression instanceof Not) {
+            if (((Not) not.expression).expression instanceof Not) {
+                return visitNot((Not) ((Not) not.expression).expression);
+            }
+        }
+        return not;
+    }
+
+    @Nonnull
+    @Override
+    protected Node visitIfElse(@Nonnull IfElse ifElse) {
+        final IfElse ifElsef = (IfElse) super.visitIfElse(ifElse);
+        Maybe<Boolean> t = Truthiness.truthiness(ifElsef.test);
+        return t.<Node>map(x -> x ? ifElsef.consequent : ifElsef.alternate).orJust(ifElsef);
+    }
+
+    @Nonnull
+    @Override
+    public NodeWithValue visitFloatMath(@NotNull FloatMath floatMath) {
+        NodeWithValue left = visitNodeWithValue(floatMath.left());
+        NodeWithValue right = visitNodeWithValue(floatMath.right());
+        if (floatMath.operator == FloatMath.Operator.Plus) {
+            if (Type(left).maybe(false, x -> x == Type.String) || Type(right).maybe(false, x -> x == Type.String)) {
+                Maybe<LiteralString> lstr = ToString(left);
+                Maybe<LiteralString> rstr = ToString(right);
+                return lstr.<NodeWithValue>flatMap(l -> rstr.map(r -> new LiteralString(l.value + r.value))).orJust(floatMath);
```

Comment:
This `orJust(floatMath)` has to be `orJustLazy(() -> new FloatMath(floatMath.operator, left, right))` like below.